<div class="alert alert-success">
    
**Комментарий ревьюера**

Привет! Меня зовут Сысоева Вера.
    
Далее в файле мои комментарии ты сможешь найти в ячейках, аналогичных данной ( если рамки комментария зелёные и отмечены **#ок** - всё сделано правильно; жёлтые и отмечены **#можно лучше** - есть замечания, но не критично; красные с пометкой **#ошибка** - нужно переделать). Не удаляй эти комментарии и постарайся учесть их в ходе выполнения проекта.
    
    
Я буду поддерживать версионность комментариев, и при следующих итерациях комментарии будут выделены, как "Комментраии ревьюера - 2 (3 и.т.д.)"
    
Если у тебя есть замечания, возражения и вопросы, ты можешь писать свои комментарии, по возможности выделяй их так же цветом (например, синим), чтоб их можно было легко отделить от текста проекта. 
    
Кстати, предлагаю сразу общаться на "ты".
    
Увидев у тебя ошибку, в первый раз я лишь укажу на ее наличие и дам тебе возможность самостоятельно найти и исправить ее. На реальной работе твой начальник будет поступать так же, а я пытаюсь подготовить тебя именно к работе аналитиком. Но если ты пока не справишься с такой задачей - при следующей проверке я дам более точную подсказку.
    
</div>

# Проект по SQL

### Цели исследования:
- исследовать базу данных, которые помогут ответить на вопросы для работа сервиса книг по подписке
- сформулировать ценностное предложение для нового продукта

### Задачи
- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Доступ к базе данных

In [13]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Исследование таблиц

In [2]:
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

**Таблица books**
Содержит данные о книгах:
book_id — идентификатор книги;
author_id — идентификатор автора;
title — название книги;
num_pages — количество страниц;
publication_date — дата публикации книги;
publisher_id — идентификатор издателя.

In [3]:
query1 = '''SELECT * FROM authors LIMIT 5'''
pd.io.sql.read_sql(sql=text(query1), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

**Таблица authors**
Содержит данные об авторах:
author_id — идентификатор автора;
author — имя автора.

In [4]:
query2 = '''SELECT * FROM ratings LIMIT 5'''
pd.io.sql.read_sql(sql=text(query2), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

**Таблица ratings**
Содержит данные о пользовательских оценках книг:
rating_id — идентификатор оценки;
book_id — идентификатор книги;
username — имя пользователя, оставившего оценку;
rating — оценка книги.

In [5]:
query3 = '''SELECT * FROM reviews LIMIT 5'''
pd.io.sql.read_sql(sql=text(query3), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

**Таблица reviews**
Содержит данные о пользовательских обзорах:
review_id — идентификатор обзора;
book_id — идентификатор книги;
username — имя автора обзора;
text — текст обзора.

In [6]:
query4 = '''SELECT * FROM publishers LIMIT 5'''
pd.io.sql.read_sql(sql=text(query4),  con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

**Таблица publishers**
Содержит данные об издательствах:
publisher_id — идентификатор издательства;
publisher — название издательства;

### Ответы на вопросы исследования

#### Вопрос 1. Посчитайте, сколько книг вышло после 1 января 2000 года.

In [7]:
task_1  =  '''SELECT  COUNT(book_id)  FROM  books  WHERE  publication_date  > '2000-01-01'  '''
task_1  =  pd.io.sql.read_sql(sql=text(task_1),  con  =  con) 
task_1


count
0    819

С начала 2000 года было выпущено 819 книг.

In [8]:
total_books  =  '''SELECT  COUNT(book_id)  FROM  books '''
total_books  =  pd.io.sql.read_sql(sql=text(total_books),  con  =  con) 
x = (task_1 / total_books)*100
x

count
0   81.9

Причем доля выпущенных книг после 1 января 2000 года составила 81.9%

#### Вопрос 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [22]:
task_2  =  '''SELECT  b.title,  AVG(r.rating),  COUNT(DISTINCT rw.review_id)  as count_reviews
FROM books b
LEFT JOIN ratings r ON b.book_id=r.book_id LEFT JOIN reviews rw ON b.book_id=rw.book_id
GROUP BY b.title ORDER  BY  avg  DESC'''
task_2  =  pd.io.sql.read_sql(sql=text(task_2),  con  =  con) 
task_2


title   avg  count_reviews
0                 Pop Goes the Weasel (Alex Cross  #5)  5.00              2
1                                          Angels Fall  5.00              2
2                 Piercing the Darkness (Darkness  #2)  5.00              2
3     The Cat in the Hat and Other Dr. Seuss Favorites  5.00              0
4                             Neil Gaiman's Neverwhere  5.00              2
..                                                 ...   ...            ...
994  The World Is Flat: A Brief History of the Twen...  2.25              3
995                                              Junky  2.00              2
996                                      Drowning Ruth  2.00              3
997                  His Excellency: George Washington  2.00              2
998                               Harvesting the Heart  1.50              2

[999 rows x 3 columns]

Таким образом есть книги, которые имеют максимальную среднюю оценку, а есть книги-аутсайдеры с минимальными оценками

<div class="alert alert-danger">
    
**Комментарий ревьюера**

**#ошибка**

По заданию, для каждой книги нужно было рассчитать показатели, т.е. в итоговой таблице должно быть 1000 строк
</div>

<div class="alert alert-info">
    
**Комментарий студента**

Сменила тип присоединения, но почему то строк 999, а не 1000
</div>

<div class="alert alert-success">
    
**Комментарий ревьюера - 2**

Да, все верно. Это потому что для одной книги есть дубликат 
</div>

#### Вопрос 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [10]:
task_3  =  '''SELECT  p.publisher,  COUNT(b.book_id)  as  total_books
FROM books b
JOIN publishers p ON b.publisher_id=p.publisher_id WHERE b.num_pages > 50
GROUP BY p.publisher ORDER BY total_books DESC LIMIT  1'''
task_3  =  pd.io.sql.read_sql(sql=text(task_3),  con  =  con) 
task_3

publisher  total_books
0  Penguin Books           42

Издательство "Penguin Books" лидирует по выпуску книг - за весь период оно успело выпустить 42 книги

#### Вопрос 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [11]:
task_4 = '''
SELECT DISTINCT a.author AS autor, AVG(r.rating) AS avg_rating FROM authors a
INNER JOIN books b ON b.author_id=a.author_id INNER JOIN ratings r ON r.book_id=b.book_id WHERE b.book_id in (
SELECT book_id FROM ratings GROUP BY book_id
HAVING COUNT(rating)>=50)
GROUP BY author
ORDER BY avg_rating DESC LIMIT 1
'''

task_4  =  pd.io.sql.read_sql(sql=text(task_4),  con  =  con) 
task_4.round(3)


autor  avg_rating
0  J.K. Rowling/Mary GrandPré       4.287

Джоа́н Ро́улинг, известная под псевдонимами Дж. К. Роулинг, Джоан Кэ́тлин Роулинг и Ро́берт Гелбре́йт, — британская писательница, сценаристка и кинопродюсер, наиболее известная как автор серии романов о Гарри Поттере. Поэтому неудивительно, что книги с ее авторством имеют наибольший рейтинг. Гарри Поттер - это книги, на которых выросло поколение!

#### Вопрос 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [12]:
task_5  =  '''SELECT
AVG(review.review_cnt) AS avg_review FROM
(SELECT COUNT(review_id) AS review_cnt FROM reviews
WHERE username IN
(SELECT username FROM ratings GROUP BY username
HAVING COUNT(ratings) > 48)
GROUP  BY  username)  AS  review'''
task_5  =  pd.io.sql.read_sql(sql=text(task_5),  con  =  con) 
task_5


avg_review
0        24.0

Таким образом, чем больше отзывов на постоянной основе ставит пользователь, тем более вероятно, что он напишет и обзор. В среднем на каждые 2 оцененные книги будет хотя бы 1 обзор у тех пользователей, которые поставили больше 48 оценок 

### Выводы и предложения

- была исследована база данных и это помогло ответить на ряд вопросов о том, какие книги нужно предлагать по подписке
- сейчас имеется всего 1000 книг, причем 819 из них достаточно новые - выпущены после 1 января 2000 года
- были выявлены книги-лидеры по отзывам и книги-аутсайдеры. необходимо уделить внимание книгам с хорошими отзывами и предлагать их наиболее активно (на них можно предлагать индивидуальную подписку)
- Чаще всего преставлены книги издательства Penguin Books - с ним можно и дальше налаживать сотрудничество на более оптимальных для нас условиях
- Наиболее хорошие отзывы у книг Дж. Роулинг - про Гарри Поттера, а это значит, что читатели любят подобную литературу
- стимулирование пользователей к оставлению отзывов может увеличить количество рецензий на книги - это тоже плюс для платформы, потому что будет легче формировать индивидуальные рекомедации

<div class="alert alert-success">
    
**Комментарий ревьюера**

**#ок**

В остальном запросы написаны верно, правильные результаты

Теперь мы знаем многие операции с `SQL` в использовании `python` и это отлично, потому что мы будем достаточно часто использовать в нашей практике.
    
Полезные ссылки по SQL:
    
https://habr.com/ru/post/268983/
    
https://habr.com/ru/post/340460/

</div>